File: packages/context-pack/src/knowledge-scope-lock.ts
import { z } from "zod";

import { canonicalJson, digestCanonicalJson } from "./canonical.js";
import type { JsonValue } from "./canonical.js";
import { BoundedJsonValueSchema } from "./bounded-json.js";
import { KnowledgeScopeDefinitionSchema } from "./knowledge-scope.js";
import type { KnowledgeScopeDefinition } from "./knowledge-scope.js";
import { PackIdSchema, SemanticVersionSchema, Sha256DigestSchema } from "./scalars.js";
import type { Sha256Digest } from "./scalars.js";

const LOCK_SCHEMA_VERSION = "knowledge-scope-lock.schift.dev/v0.1" as const;
const PortablePathSchema = z.string().min(1).refine((path) =>
  !path.startsWith("/") && !path.includes("\\") && !path.includes("://") &&
  path.split("/").every((component) => component.length > 0 && component !== "." && component !== ".."),
"Lock paths must be normalized relative paths");
const LockFileSchema = z.object({ path: PortablePathSchema, digest: Sha256DigestSchema }).strict().readonly();
const LockProjectionBaseSchema = z.object({
  schemaVersion: z.literal(LOCK_SCHEMA_VERSION), packId: PackIdSchema,
  version: SemanticVersionSchema, definitionDigest: Sha256DigestSchema,
  files: z.array(LockFileSchema).min(1).readonly(),
}).strict();
const validateSortedFiles = (lock: z.infer<typeof LockProjectionBaseSchema>, context: z.RefinementCtx): void => {
  const paths = lock.files.map((file) => file.path);
  if (new Set(paths).size !== paths.length || canonicalJson(paths) !== canonicalJson([...paths].sort())) {
    context.addIssue({ code: z.ZodIssueCode.custom, message: "Lock files must be unique and lexically sorted" });
  }
};
const LockProjectionSchema = LockProjectionBaseSchema.superRefine(validateSortedFiles).readonly();
export const KnowledgeScopeLockSchema = LockProjectionBaseSchema.extend({
  lockDigest: Sha256DigestSchema,
}).strict().superRefine(validateSortedFiles).readonly();

export type KnowledgeScopeLock = z.infer<typeof KnowledgeScopeLockSchema>;
export type KnowledgeScopeFileValues = Readonly<Record<string, JsonValue>>;

export class KnowledgeScopeLockError extends Error {
  public readonly code: "invalid_file_inventory" | "definition_file_mismatch" | "non_portable_number";

  public constructor(code: KnowledgeScopeLockError["code"], message: string) {
    super(message);
    this.name = "KnowledgeScopeLockError";
    this.code = code;
  }
}

const normalizePortableJson = (value: JsonValue): JsonValue => {
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new KnowledgeScopeLockError(
        "non_portable_number",
        "Knowledge Scope lock JSON numbers must be portable safe integers",
      );
    }
    return value;
  }
  if (Array.isArray(value)) return value.map(normalizePortableJson);
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, entry]) => [key, normalizePortableJson(entry)]),
    );
  }
  return value;
};

export const digestKnowledgeScopeDefinition = async (
  definition: KnowledgeScopeDefinition,
): Promise<Sha256Digest> => digestCanonicalJson(
  normalizePortableJson(BoundedJsonValueSchema.parse(definition)),
);

const expectedPaths = (definition: KnowledgeScopeDefinition): readonly string[] => [
  "scope.json",
  ...new Set(definition.capabilities.flatMap((capability) => [
    capability.inputSchemaRef, capability.resultSchemaRef,
  ])),
].sort();

const validateInventory = (
  definition: KnowledgeScopeDefinition,
  files: KnowledgeScopeFileValues,
): readonly string[] => {
  const paths = Object.keys(files).sort();
  const parsed = z.array(PortablePathSchema).safeParse(paths);
  if (!parsed.success || canonicalJson(paths) !== canonicalJson(expectedPaths(definition))) {
    throw new KnowledgeScopeLockError("invalid_file_inventory", "Files must exactly match scope.json and declared schemas");
  }
  const scopeFile = files["scope.json"];
  const parsedScope = KnowledgeScopeDefinitionSchema.safeParse(scopeFile);
  if (!parsedScope.success ||
    canonicalJson(normalizePortableJson(BoundedJsonValueSchema.parse(parsedScope.data))) !==
      canonicalJson(normalizePortableJson(BoundedJsonValueSchema.parse(definition)))) {
    throw new KnowledgeScopeLockError("definition_file_mismatch", "scope.json must match the parsed definition");
  }
  return paths;
};

export const buildKnowledgeScopeLock = async (
  definition: KnowledgeScopeDefinition,
  files: KnowledgeScopeFileValues,
): Promise<KnowledgeScopeLock> => {
  const paths = validateInventory(definition, files);
  const lockProjection = LockProjectionSchema.parse({
    schemaVersion: LOCK_SCHEMA_VERSION,
    packId: definition.packId,
    version: definition.version,
    definitionDigest: await digestKnowledgeScopeDefinition(definition),
    files: await Promise.all(paths.map(async (path) => ({
      path,
      digest: await digestCanonicalJson(
        normalizePortableJson(BoundedJsonValueSchema.parse(files[path] ?? null)),
      ),
    }))),
  });
  return KnowledgeScopeLockSchema.parse({
    ...lockProjection, lockDigest: await digestCanonicalJson(lockProjection),
  });
};

export const verifyKnowledgeScopeLock = async (
  definition: KnowledgeScopeDefinition,
  files: KnowledgeScopeFileValues,
  lock: KnowledgeScopeLock,
): Promise<boolean> => {
  const parsed = KnowledgeScopeLockSchema.safeParse(lock);
  if (!parsed.success) return false;
  const expected = await buildKnowledgeScopeLock(definition, files);
  return canonicalJson(expected) === canonicalJson(parsed.data);
};

File: packages/core/context-pack/python/schift_context_pack_contract/knowledge_scope_lock.py
"""Deterministic portable lock for a Knowledge Scope definition."""

# Recursive JSON matching keeps a future-proof exhaustive default.
# pyright: reportUnnecessaryComparison=false

from __future__ import annotations

import math
from collections.abc import Mapping
from typing import Final, Literal, assert_never, final, override

from pydantic import Field

from schift_context_pack_contract.json_boundary import JsonValue, canonical_digest, to_json_value
from schift_context_pack_contract.knowledge_scope import KnowledgeScopeDefinition
from schift_context_pack_contract.knowledge_scope_digest import knowledge_scope_definition_digest
from schift_context_pack_contract.manifest import (
    ContractModel,
    Identifier,
    PortablePath,
    SemVer,
    Sha256Digest,
)

type KnowledgeScopeJsonFiles = Mapping[str, JsonValue]
_MAX_SAFE_INTEGER: Final = 2**53 - 1


@final
class KnowledgeScopeLockError(Exception):
    __slots__ = ("code", "path")

    def __init__(self, *, code: str, path: str) -> None:
        super().__init__()
        self.code = code
        self.path = path

    @override
    def __str__(self) -> str:
        return f"{self.code}: {self.path}"


class KnowledgeScopeLockFile(ContractModel):
    path: PortablePath
    digest: Sha256Digest


class KnowledgeScopeLock(ContractModel):
    schema_version: Literal["knowledge-scope-lock.schift.dev/v0.1"] = Field(alias="schemaVersion")
    pack_id: Identifier = Field(alias="packId")
    version: SemVer
    definition_digest: Sha256Digest = Field(alias="definitionDigest")
    files: tuple[KnowledgeScopeLockFile, ...]
    lock_digest: Sha256Digest = Field(alias="lockDigest")


def _required_paths(definition: KnowledgeScopeDefinition) -> tuple[str, ...]:
    paths = {"scope.json"}
    for capability in definition.capabilities:
        paths.add(capability.input_schema_ref)
        paths.add(capability.result_schema_ref)
    return tuple(sorted(paths))


def _validate_inventory(
    definition: KnowledgeScopeDefinition,
    files: KnowledgeScopeJsonFiles,
) -> tuple[str, ...]:
    required = _required_paths(definition)
    supplied = tuple(sorted(files))
    if supplied != required:
        raise KnowledgeScopeLockError(code="invalid_file_inventory", path=",".join(supplied))
    for path in supplied:
        parts = path.split("/")
        if path.startswith("/") or "\\" in path or any(part in {"", ".", ".."} for part in parts):
            raise KnowledgeScopeLockError(code="invalid_portable_path", path=path)
    scope = KnowledgeScopeDefinition.model_validate(files["scope.json"])
    if scope != definition:
        raise KnowledgeScopeLockError(code="definition_file_mismatch", path="scope.json")
    return supplied


def _lock_projection(lock: KnowledgeScopeLock) -> JsonValue:
    return to_json_value(lock.model_dump_json(by_alias=True, exclude={"lock_digest"}))


def _normalize_lock_value(value: JsonValue, path: str) -> JsonValue:
    match value:
        case dict() as mapping:
            return {
                key: _normalize_lock_value(child, f"{path}.{key}") for key, child in mapping.items()
            }
        case list() as items:
            return [
                _normalize_lock_value(child, f"{path}[{index}]")
                for index, child in enumerate(items)
            ]
        case bool() | str() | None:
            return value
        case int() as number:
            if abs(number) > _MAX_SAFE_INTEGER:
                raise KnowledgeScopeLockError(code="non_portable_number", path=path)
            return number
        case float() as number:
            if (
                not math.isfinite(number)
                or not number.is_integer()
                or abs(number) > _MAX_SAFE_INTEGER
            ):
                raise KnowledgeScopeLockError(code="non_portable_number", path=path)
            return int(number)
        case unreachable:
            assert_never(unreachable)


def build_knowledge_scope_lock(
    definition: KnowledgeScopeDefinition,
    files: KnowledgeScopeJsonFiles,
) -> KnowledgeScopeLock:
    """Build the canonical definition-only lock over exact parsed JSON files."""
    paths = _validate_inventory(definition, files)
    entries = tuple(
        KnowledgeScopeLockFile(
            path=path,
            digest=canonical_digest(_normalize_lock_value(files[path], path)),
        )
        for path in paths
    )
    unsigned = {
        "schemaVersion": "knowledge-scope-lock.schift.dev/v0.1",
        "packId": definition.pack_id,
        "version": definition.version,
        "definitionDigest": knowledge_scope_definition_digest(definition),
        "files": tuple(item.model_dump(by_alias=True) for item in entries),
        "lockDigest": "sha256:" + ("0" * 64),
    }
    draft = KnowledgeScopeLock.model_validate(unsigned)
    return draft.model_copy(update={"lock_digest": canonical_digest(_lock_projection(draft))})


def verify_knowledge_scope_lock(
    definition: KnowledgeScopeDefinition,
    files: KnowledgeScopeJsonFiles,
    lock: KnowledgeScopeLock,
) -> bool:
    """Return whether a lock exactly matches its definition and JSON inventory."""
    try:
        expected = build_knowledge_scope_lock(definition, files)
    except KnowledgeScopeLockError:
        return False
    return expected == lock
